review/ark-blockstore-cstor/cstor.go

```diff
@@ -0,0 +1,526 @@
+                networkIp, ok := netInterfaceAddress.(*net.IPNet)
+                if ok && !networkIp.IP.IsLoopback() && networkIp.IP.To4() != nil {
+                        ip := networkIp.IP.String()
+                        logrus.Infof("Resolved Host IP: " + ip)
+                        return ip + ":" + strconv.Itoa(RecieverPort)
+                }
+        }
+        return ""
+}
+
+func (p *cstorSnapPlugin) getMapiAddr() string {
+	sc, err := p.K8sClient.Services(operator).Get(mayaAPIServiceName, metav1.GetOptions{})
+        if err != nil {
+                p.Log.Errorf("Error getting IP Address for service - %s : %v", mayaAPIServiceName, err)
+		return ""
+        }
+
+        if len(sc.Spec.ClusterIP) != 0 {
+		return "http://" + sc.Spec.ClusterIP + ":" + strconv.FormatInt(int64(sc.Spec.Ports[0].Port), 10)
+        } else {
+		return ""
+	}
+}
+
+
+func (p *cstorSnapPlugin) Init(config map[string]string) error {
+        conf, err := rest.InClusterConfig()
+        if err != nil {
+		p.Log.Errorf("Failed to get cluster config", err)
+		return errors.New("Error fetching cluster config")
+        }
+
+	clientset, err := kubernetes.NewForConfig(conf)
+        if err != nil {
+		p.Log.Errorf("Error creating clientset", err)
+		return errors.New("Error creating k8s client")
+        }
+
+	p.K8sClient = clientset.CoreV1()
+	p.mayaAddr = p.getMapiAddr()
+	if p.mayaAddr == "" {
+		return errors.New("Error fetching OpenEBS rest client address")
+	}
+
+	p.cstorServerAddr = p.getServerAddress()
+	if p.cstorServerAddr == "" {
+		return errors.New("Error fetch cstorArkServer address")
+	}
+	p.config = config
+	if p.volumes == nil {
+		p.volumes = make(map[string]*Volume)
+	}
+	if p.snapshots == nil {
+		p.snapshots = make(map[string]*Snapshot)
+	}
+
+	p.cl = &cloudUtils {Log: p.Log}
+	return p.cl.InitCloudConn(config)
+}
+
+func (p *cstorSnapPlugin) GetVolumeID(pv runtime.Unstructured) (string, error) {
+        if !collections.Exists(pv.UnstructuredContent(), "metadata") {
+                return "", nil
+        }
+
+        // Seed the volume info so that GetVolumeInfo doesn't fail later.
+        volumeID, _ := collections.GetString(pv.UnstructuredContent(), "metadata.name")
+        if _, exists := p.volumes[volumeID]; !exists {
+                sc, _ := collections.GetString(pv.UnstructuredContent(), "spec.storageClassName")
+                ns, _ := collections.GetString(pv.UnstructuredContent(), "spec.claimRef.namespace")
+                p.volumes[volumeID] = &Volume{
+			volname: volumeID,
+                        casType: sc,
+                        namespace: ns,
+                }
+        }
+
+        return collections.GetString(pv.UnstructuredContent(), "metadata.name")
+}
+
+func (p *cstorSnapPlugin) DeleteSnapshot(snapshotID string) error {
+        var snapInfo *Snapshot
+        var err error
+
+        p.Log.Infof("Deleting snapshot %v", snapshotID)
+        if _, exists := p.snapshots[snapshotID]; !exists {
```

Comment:
Does it store a list of snapshots in-memory?
If there are thousands of snapshots?